apps/merkle_patricia_tree/lib/trie.ex

```diff
@@ -77,96 +76,118 @@ defmodule MerklePatriciaTree.Trie do
       :empty ->
         nil
 
-      {:branch, branches} ->
-        # branch node
-        case Enum.at(branches, nibble) do
-          [] -> nil
-          node_hash -> node_hash |> into(trie) |> do_get(rest)
-        end
-
+      # Leaf node
       {:leaf, prefix, value} ->
-        # leaf, value is second value if match first
-        case nibbles do
-          ^prefix -> value
-          _ -> nil
-        end
+        if prefix == nibbles,
+          do: value,
+          else: nil
 
+      # Extension, continue walking trie if we match
       {:ext, shared_prefix, next_node} ->
-        # extension, continue walking tree if we match
         case ListHelper.get_postfix(nibbles, shared_prefix) do
-          # did not match extension node
+          # Did not match extension node
           nil ->
             nil
 
           rest ->
             next_node |> into(trie) |> do_get(rest)
         end
+
+      # Branch node
+      {:branch, branches} ->
+        case Enum.at(branches, nibble) do
+          [] -> nil
+          node_hash -> node_hash |> into(trie) |> do_get(rest)
+        end
     end
   end
 
   defp do_get(trie, []) do
-    # Only branch nodes can have values for a nil lookup
+    # No prefix left, its either branch or leaf node
     case Node.decode_trie(trie) do
-      {:branch, branches} -> List.last(branches)
-      {:leaf, [], v} -> v
-      _ -> nil
+      # In branch node value is always the last element
+      {:branch, branches} ->
+        List.last(branches)
+
+      {:leaf, [], v} ->
+        v
+
+      _ ->
+        nil
     end
   end
 
   @doc """
   Updates a trie by setting key equal to value. If value is nil,
   we will instead remove `key` from the trie.
   """
-  @spec update(__MODULE__.t(), __MODULE__.key(), ExRLP.t() | nil) :: __MODULE__.t()
-  def update(trie, key, nil) do
-    trie
-    |> Node.decode_trie()
-    |> Destroyer.remove_key(Helper.get_nibbles(key), trie)
-    |> Node.encode_node(trie)
-    |> into(trie)
-    |> store
-  end
+  @spec update(t(), key(), ExRLP.t() | nil) :: t()
+  def update(trie, key, nil), do: remove(trie, key)
 
   def update(trie, key, value) do
+    key_nibbles = Helper.get_nibbles(key)
     # We're going to recursively walk toward our key,
     # then we'll add our value (either a new leaf or the value
     # on a branch node), then we'll walk back up the tree and
     # update all previous nodes.
     # This may require changing the type of the node.
     trie
     |> Node.decode_trie()
-    |> Builder.put_key(Helper.get_nibbles(key), value, trie)
+    |> Builder.put_key(key_nibbles, value, trie)
     |> Node.encode_node(trie)
     |> into(trie)
     |> store
   end
 
-  def store(trie) do
-    root_hash =
-      if not is_binary(trie.root_hash) or trie.root_hash == <<>>,
-        do: ExRLP.encode(trie.root_hash),
-        else: trie.root_hash
+  @doc """
+  Removes `key` from the `trie`.
+  """
+  @spec remove(t(), key()) :: t()
+  def remove(trie, key) do
```

Comment:
Do you think `remove_key` is a better name?

Reply:
I've named it `remove` for consistency with the above function `update`.
It tells the "truth" because it also removes the value.